### ISSUE TITLE
Bug: [strict-boolean-expressions] False positive with union type containing allowed falsy types

### ISSUE THREAD

**Issue by chronoDave** [typescript-eslint/typescript-eslint#4969]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.6.4&sourceType=module&code=JYWwDg9gTgLgBAMwM6KhEcDkzMG4CwAUEQMYQB2S8AhgFyJIB0ACtTABYAywA1gKZwAPnACu5ACZ8Ewcn3FwA9ArhUoMgOZC4AIREIEfKFoCqAJU4FCZSvABG9VRq1jJ02eMvWqcEvWMyYAA4AQSgoagBPZwkpGTlPCm9xP3No1ziPIiJgBDgACgBCagBKH0SIABs+RgqIdTzMAFEADzA+Ehg5OHIIOEM0KExiyxz8gttSr0rq2vqm1vbO+R6+sOghkdzCkknyqpq6hpa2jq6V-vXh7K2C8V3KaYO548Wz3ovB4qIgA&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AhOktgEYD2NucAhgTgB4wqIJo0EIgAuANog4KFDRQgANKEa5cNAO4A5AK4LGVBgGVyhAOaDyauAF8AumaA&tsConfig=N4Igpgzgsg9gJgVwDZgJIDsAuYBOMAOIAXJjgmADQgSkCWAxpgHLJIDCAFmPQNYTGlyAXyA

### Repro Code

```typescript
import fs from 'fs';

const a: fs.PathLike | undefined // string | Buffer | URL;
const b: string | undefined;
const c: Uint8Array | undefined;
const d: URL | undefined;

if (!a) console.log('Expected no error');
if (!b) console.log('Expected no error');
if (!c) console.log('Expected no error');
if (!d) console.log('Expected no error')
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/strict-boolean-expressions": [
      "error",
      {
        "allowNullableString": true
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "esModuleInterop": true,
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected the 3rd line to not throw an error. None of the individual checks (`b`, `c`, `d`) throw an error, but the combined type (`a`) does.

### Actual Result

There was an error on the 3rd line:

`Unexpected any value in conditional. An explicit comparison or type cast is required.ESLint(@typescript-eslint/strict-boolean-expressions)`

### Additional Info

I've tried looking for similar issues but I'm not sure if these issues are related:

- https://github.com/typescript-eslint/typescript-eslint/issues/2894
- https://github.com/typescript-eslint/typescript-eslint/issues/3115

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `4.33.0` |
| `@typescript-eslint/parser`            | `4.33.0` |
| `TypeScript`                           | `4.6.4` |
| `ESLint`                               | `7.32.0` |
| `node`                                 | `16.13.0` |

**Comment by bradzacher** [comment]
The type of `fs.PathLike` is `string | Buffer | URL`. The latter two are classes, so the type of `a` is `string | object | undefined`

This whilst this type is made up of a "nullable object" and a "nullable string", as a whole is is neither - it is a "nullable object or string". This is why it errors, and this is entirely on purpose.

Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/2894
